Implement card view for tasks view of a dag

tirkarthi opened the pull request:
Use `tasks` endpoint to fill the card view first. Get dag's recent dagruns and get the last dagrun id and pass it to taskinstances endpoint to get the last run and the corresponding task instance state. 

Comments for reviewers and self :

1. Disable pagination since tasks and task instances doesn't support limit or pagination.
2. Colors are slightly off with `stateColor` and the theme which will be fixed overall. Please let me know if I am missing some other configuration.
3. The historical view of the task instances which is the last column in the card view is using list task instances endpoint which returns like 100 task instances and the limit is easily reached with 14 task instances per task. All task instance attributes are also not used with only start_date, end_date, try_number, duration and state used which could be fetched efficiently.
4. Do we need to display a tooltip for the last run. Perhaps the start time, end time and try number of the last task instance.

This PR probably needs a rebase after https://github.com/apache/airflow/pull/44269 is merged.

![image](https://github.com/user-attachments/assets/71d01a4a-93af-43bf-8b63-b6f2886293a9)

![image](https://github.com/user-attachments/assets/adbe71c5-57aa-401b-8db7-c2f5389e04e7)

tirkarthi committed:
Initial commit for tasks tab.

tirkarthi committed:
Remove unused columns for card view.

tirkarthi committed:
Fetch latest dag run and then fetch corresponding task instances.

tirkarthi committed:
Make font to be consistent.

tirkarthi committed:
Pass dagRunsLimit to limit the number of dagruns.

bbovenzi commented on airflow/ui/src/pages/DagsList/Dag/Tasks/TaskCard.tsx:
```diff
@@ -0,0 +1,83 @@
+            <Text> {taskInstance[0].logical_date} </Text>
+            {taskInstance[0].state === null ? undefined : (
+              <>
+                <StateCircle state={taskInstance[0].state} />
```
Once the runs PR is merged we can swap this for `<Status>` and we'll deviate from the designs by not making the time component have the state's color

tirkarthi replied:
Fixed conflicts and used `<Status />` since runs PR is merged. Thanks.

bbovenzi commented on airflow/ui/src/pages/DagsList/Dag/Tasks/Tasks.tsx:
```diff
@@ -0,0 +1,112 @@
+    useTaskInstanceServiceGetTaskInstances(
+      {
+        dagId: dagId ?? "",
+        dagRunId: runs[0]?.dag_run_id ?? "~",
```
I think it would be great to get the last 14 TIs and make a bar chart similar to what we have for the Dag Card

tirkarthi replied:
I left it intentionally for another PR since that I guess will make an API call per dagrun to fetch task instances and I felt querying a lot of data like rendered templates to show only the duration field that was not efficient. Probably needs a `task_stats` or `grid_data` like endpoint that is specialized for this.

tirkarthi replied:
I implemented is similar to `RecentRuns` but the current public endpoint for task instances returns only 100 task instances which is limited for the view. I have added a TODO for this though the UI functionality is there and will come back later once API is available for this. I have updated the screenshots. Thanks.

tirkarthi replied:
I am working on an API in another branch to filter by dag_id and optionally state to return the recent 14 task instances per task. Following is draft response and I will update this page and also runs page to use this that will be more efficient. Since this private ui endpoint this can be later updated as per needs without concern for backwards compatibility,

```
http :8000/ui/task_instances/recent_task_instances dag_id==example_complex state==failed
```

```json
{
    "task_instances": [
        {
            "dag_id": "example_complex",
            "end_date": "2024-12-04T10:18:42.969547Z",
            "start_date": "2024-12-04T10:18:42.779151Z",
            "state": "failed",
            "task_id": "create_entry_gcs",
            "try_number": 1
        },
    ],
    "total_entries": 1
}
```

bbovenzi replied:
Looks good! Let's not forget `dag_run_id` too

bbovenzi commented on airflow/ui/src/pages/DagsList/Dag/Tasks/Tasks.tsx:
```diff
@@ -0,0 +1,112 @@
+        isFetching={isFetching}
+        isLoading={isLoading}
+        modelName="Task"
+        total={data ? data.total_entries : 0} // Todo : Disable pagination?
```
Good call out. We should update DataTable to only enable manualPagination and manualSorting if there is are `initialState` and `handleStateChange` props. And then only show the pagination component if there is manualPagination.

tirkarthi replied:
My concern was around some dags in real world having hundred or so tasks that will increase rendering time but the current fetch dag tasks API doesn't support pagination.

tirkarthi committed:
Add recent task instance plot to card.

tirkarthi committed:
Fix merge conflicts and use Status component.

bbovenzi commented on airflow/ui/src/pages/DagsList/Dag/Tasks/TaskCard.tsx:
```diff
@@ -0,0 +1,84 @@
+    borderWidth={1}
+    overflow="hidden"
+  >
+    <Text bg="bg.info" color="fg.info" p={2}>
```
```suggestion
    <Text bg="bg.info" color="fg.info" fontWeight="bold" p={2}>
```

tirkarthi replied:
The design had `bg` with info color. Do we need to retain `bg="bg.info"` ?

bbovenzi replied:
You're right, my bad. Let's just make the text bold.

tirkarthi replied:
Thanks, done.

bbovenzi commented on airflow/ui/src/pages/DagsList/Dag/Tasks/TaskCard.tsx:
```diff
@@ -0,0 +1,84 @@
+      </VStack>
+      <VStack align="flex-start" gap={1}>
+        <Heading color="fg.muted" fontSize="xs">
+          Last Run
```
```suggestion
          Last Instance
```

tirkarthi replied:
Thanks, done.

bbovenzi commented on airflow/ui/src/pages/DagsList/Dag/Tasks/TaskCard.tsx:
```diff
@@ -0,0 +1,84 @@
+                {taskInstances[0].state}
+              </Status>
+            )}
+          </HStack>
```
Let's wrap this in a tooltip like we do with the latest dag run?
Include the dag run id and state, and the TI start/end/try_number

tirkarthi replied:
Thanks, added those fields and moved it to a separate component as TaskInstanceTooltip to be used in Last Instance field and also the bar chart. It seems I had to wrap the element and pass it as a child. Please let me know if there is a better convention to do it in React.

bbovenzi commented on airflow/ui/src/pages/DagsList/Dag/Tasks/TaskRecentRuns.tsx:
```diff
@@ -0,0 +1,97 @@
+                <Text>
+                  End Date: <Time datetime={taskInstance.end_date} />
+                </Text>
+                <Text>Try Number: {taskInstance.try_number}</Text>
```
Let's only show try_number if it's greater than 1

tirkarthi replied:
Thanks, done.

bbovenzi commented on airflow/ui/src/pages/DagsList/Dag/Tasks/TaskRecentRuns.tsx:
```diff
@@ -0,0 +1,97 @@
+          <Tooltip
+            content={
+              <Box>
+                <Text>State: {taskInstance.state}</Text>
```
Let's include the dag run id and state too.

We could probably make this tooltip into its own `TaskInstanceTooltip` and use it for the "Latest Instance" card field too.

tirkarthi replied:
Thanks, used `TaskInstanceTooltip` component.

bbovenzi commented on airflow/ui/src/pages/DagsList/Dag/Tasks/Tasks.tsx:
```diff
@@ -0,0 +1,113 @@
+    runsData?.dags.find((dagWithRuns) => dagWithRuns.dag_id === dagId)
+      ?.latest_dag_runs ?? [];
+
+  // TODO: Revisit this endpoint since only 100 task instances are returned and
+  // only duration is calculated with other attributes unused.
```
Yes good point. We'll need a custom ui endpoint in the API to properly power this. At the bare minimum the public endpoint should have a `task_display_name_pattern` param.

bbovenzi commented on airflow/ui/src/pages/DagsList/Dag/Tasks/Tasks.tsx:
```diff
@@ -0,0 +1,113 @@
+    <Box>
+      <ErrorAlert error={TasksError} />
+      <Heading my={1} size="md">
+        {data ? data.total_entries : 0} Tasks
```
Let's use `pluralize()` here

tirkarthi replied:
Thanks, done

bbovenzi commented on airflow/ui/src/router.tsx:
```diff
@@ -26,8 +26,10 @@ import { Overview } from "src/pages/DagsList/Dag/Overview";
+
+import { Tasks } from "./pages/DagsList/Dag/Tasks";
+import { ErrorPage } from "./pages/Error";
+import { Events } from "./pages/Events";
```
```suggestion
import { Tasks } from "src/pages/DagsList/Dag/Tasks";
import { ErrorPage } from "src/pages/Error";
import { Events } from "src/pages/Events";
```

I might open a PR to add more aliases so that we can just call `pages/Events`

tirkarthi replied:
Thanks, done. I guess I did this during rebase after runs PR merge.

bbovenzi commented on airflow/ui/src/pages/DagsList/Dag/Tasks/TaskCard.tsx:
```diff
@@ -0,0 +1,84 @@
+        </Heading>
+        {taskInstances[0] ? (
+          <HStack fontSize="sm">
+            <Text> {taskInstances[0].logical_date} </Text>
```
This should be inside of a `<Time>` component. And I think we want to use start_date here.

tirkarthi replied:
Thanks, switched to start_date

tirkarthi committed:
Make task name bold
Change Last Run to Last Instance in header name
Use start date and wrap it with Time tag
Show try number only when greater than 1
Use pluralize for task

bbovenzi commented on airflow/ui/src/pages/DagsList/Dag/Tasks/Tasks.tsx:
```diff
@@ -0,0 +1,113 @@
+
+  // TODO: Revisit this endpoint since only 100 task instances are returned and
+  // only duration is calculated with other attributes unused.
+  const { data: TaskInstancesResponse } =
```
```suggestion
  const { data: taskInstancesResponse } =
```

This is not a component or type. The name should start with a lowercase letter

tirkarthi replied:
Thanks, done. Will try to keep in mind in future PRs too.

bbovenzi commented on airflow/ui/src/pages/DagsList/Dag/Tasks/Tasks.tsx:
```diff
@@ -0,0 +1,113 @@
+  const { dagId } = useParams();
+  const {
+    data,
+    error: TasksError,
```
```suggestion
    error: tasksError,
```

tirkarthi committed:
Refactor tooltip to a separate component. Fix variable name casing.

tirkarthi committed:
Fix duration when null.

bbovenzi commented on airflow/ui/src/components/TaskInstanceTooltip.tsx:
```diff
@@ -0,0 +1,64 @@
+import type { TaskInstanceResponse } from "openapi/requests/types.gen";
+import Time from "src/components/Time";
+import { Tooltip } from "src/components/ui";
+
```
Let's define props like so:
```
type Props = {
  readonly taskInstance: TaskInstanceResponse
} & TooltipProps;

Then we can use `children` instead of `child` and just wrap this tooltip around any component like `<TaskInstanceTooltip taskInstance={ti}><MyJsxComponent /></TaskInstanceTooltip>`

tirkarthi replied:
Thanks TIL about pattern using `props.children` It seems `content` is required in `Tooltip` which I construct but using `Props & TooltipProps` causes typescript to expect `content` to be passed. I declared `children` explicitly inside `Props` and passed it on.

```
src/pages/DagsList/Dag/Tasks/TaskCard.tsx:72:12 - error TS2741: Property 'content' is missing in type '{ children: Element; taskInstance: TaskInstanceResponse; }' but required in type '{ content: ReactNode; contentProps?: TooltipContentProps | undefined; disabled?: boolean | undefined; portalled?: boolean | undefined; portalRef?: RefObject<...> | undefined; showArrow?: boolean | undefined; }'.

72           <TaskInstanceTooltip taskInstance={taskInstances[0]}>
              ~~~~~~~~~~~~~~~~~~~

  src/components/ui/Tooltip.tsx:23:3
    23   content: React.ReactNode;
         ~~~~~~~
    'content' is declared here.
```

bbovenzi replied:
Oh we can omit content `& Omit<TooltipProps, "content">`

tirkarthi replied:
Thanks, done.

bbovenzi commented on airflow/ui/src/components/TaskInstanceTooltip.tsx:
```diff
@@ -0,0 +1,64 @@
+
+import type { TaskInstanceResponse } from "openapi/requests/types.gen";
+import Time from "src/components/Time";
+import { Tooltip } from "src/components/ui";
```
```suggestion
import { Tooltip, type TooltipProps } from "src/components/ui";
```

tirkarthi committed:
Use children prop to simplify TaskInstanceTooltip wrapping.

tirkarthi committed:
Omit content to reuse tooltip type.
